server/api/src/rs_get_rescinfo_for_put.cpp

```diff
@@ -0,0 +1,83 @@
+#include <cstring> // for strdup.
+#include <nlohmann/json.hpp> // for strdup.
+
+//// ----------- TODO: make this file
+// #include "irods/rs_get_rescinfo_for_put.hpp"
+
+#include "irods/catalog_utilities.hpp"
+#include "irods/irods_logger.hpp"
+#include "irods/rodsErrorTable.h"
+
+#include "irods/getHostForPut.h"
+#include "irods/rodsLog.h"
+#include "irods/rsGlobalExtern.hpp"
+#include "irods/rcGlobalExtern.h"
+#include "irods/getRemoteZoneResc.h"
+#include "irods/dataObjCreate.h"
+#include "irods/objMetaOpr.hpp"
+#include "irods/resource.hpp"
+#include "irods/collection.hpp"
+#include "irods/specColl.hpp"
+#include "irods/miscServerFunct.hpp"
+#include "irods/rsGetHostForPut.hpp"
+
+// =-=-=-=-=-=-=-
+#include "irods/irods_resource_backport.hpp"
+#include "irods/irods_resource_redirect.hpp"
+
+
+namespace
+{
+    using log_api = irods::experimental::log::api;
+} // anonymous namespace
+
+
+auto rs_get_rescinfo_for_put(RsComm* rsComm, dataObjInp_t *dataObjInp, char** _resp) -> int
+{
+    rodsServerHost_t *rodsServerHost;
+    const int remoteFlag = getAndConnRemoteZone(rsComm, dataObjInp, &rodsServerHost, REMOTE_OPEN);
+    if (remoteFlag < 0) {
+        return remoteFlag;
+    }
+    else if (REMOTE_HOST == remoteFlag) {
+        const int status = rc_get_rescinfo_for_put(rodsServerHost->conn, dataObjInp, _resp);
+        if (status < 0) {
+            return status;
+        }
+    }
+    else {
+        // =-=-=-=-=-=-=-
+        // working on the "home zone", determine if we need to redirect to a different
+        // server in this zone for this operation.  if there is a RESC_HIER_STR_KW then
+        // we know that the redirection decision has already been made
+        std::string hier{};
+        if ( getValByKey( &dataObjInp->condInput, RESC_HIER_STR_KW ) == NULL ) {
+            try {
+                auto result = irods::resolve_resource_hierarchy(irods::CREATE_OPERATION, rsComm, *dataObjInp);
```

Comment:
Let's discuss during standup.

Reply:
or we move to a new thing that is well named and wrap the old behavior/call?

Reply:
Would the parameter take the form of a new or existing keyword ?  Like `RESOURCE_INFO_KW`

Reply:
that doesn't seem unreasonable.   discush!

Reply:
Post discussion ideas ...

API name suggestions:
- `rx_get_resource_info_for_operation` **<-- winner (for now)**
- `rx_get_resource_info_for_transfer`
- `rx_get_rescinfo_for_operation`
- `rx_get_rescinfo_for_transfer`

Conditional Input (i.e. `DataObjInp::condInput`):
- `OPERATION_TYPE_KW` / `OPR_TYPE_KW`
    - Allows the caller to change the operation type passed to `resolve_resource_hierarchy`
    - These keywords may be too generic
- `GET_RESOURCE_INFO_OP_TYPE_KW` **<-- winner (for now)**
    - Valid values for this keyword:
        - `"CREATE"`
        - `"WRITE"`
        - `"OPEN"`
        - `"UNLINK"`

Reply:
i like it.  both current winners work for me.

API... do we need the 'info'?   `rx_get_resource_for_operation` ?

which would make the kw name... `GET_RESOURCE_OP_TYPE_KW` ?

not wedded to these two ideas.

Reply:
`info` was included because we get a JSON payload back. That payload can grow more information over time.

Currently, it includes the hostname of the resource and its hierarchy.

Reply:
certainly very reasonable.

Reply:
This has been done, but will not resolve yet, so that the convo remains conspicuous.